lms/djangoapps/certificates/api.py

```diff
@@ -0,0 +1,118 @@
+@require_POST
+def generate_user_cert(request, course_id):
+    """
+    It will add the add-cert request into the xqueue.
+
+     Arguments:
+        request (django request object):  the HTTP request object that triggered this view function
+        course_id (unicode):  id associated with the course
+
+    Returns:
+        returns json response
+    """
+
+    if not request.user.is_authenticated():
+        log.info(u"Anon user trying to generate certificate for %s", course_id)
+        return JsonResponseBadRequest(_('You must be logged-in to generate certificate'))
+
+    student = request.user
+
+    # checking course id
+    try:
+        course_key = CourseKey.from_string(course_id)
+    except InvalidKeyError:
+        return JsonResponseBadRequest(_("Course Id is not valid"))
+
+    course = modulestore().get_course(course_key, depth=2)
+    if not course:
+        return JsonResponseBadRequest(_("Course is not valid"))
+
+    if not is_course_passed(course, None, student, request):
+        return JsonResponseBadRequest(_("You failed to pass the course."))
+
+    xqueue = XQueueCertInterface()
+
+    certificate_status = certificate_downloadable_status(student, course_key)
+    if not certificate_status["is_downloadable"] and not certificate_status["is_generating"]:
+        ret = xqueue.add_cert(student, course_key, course=course)
+        log.info(
+            (
+                u"Added a certificate generation task to the XQueue "
+                u"for student %s in course '%s'. "
+                u"The new certificate status is '%s'."
+            ),
+            student.id,
+            unicode(course_key),
+            ret
+        )
+        return JsonResponse(_("Certificate generated."))
+        # for any other status return bad request response
+    return JsonResponseBadRequest('')
+
+
+def is_course_passed(course, grade_summary=None, student=None, request=None):
+    """
+    check user's course passing status. return True if passed
+
+    Arguments:
+        course : course object
+        grade_summary (dict) : contains student grade details.
+        student : user object
+        request (HttpRequest)
+
+    Returns:
+        returns bool value
+    """
+    nonzero_cutoffs = [cutoff for cutoff in course.grade_cutoffs.values() if cutoff > 0]
+    success_cutoff = min(nonzero_cutoffs) if nonzero_cutoffs else None
+
+    if grade_summary is None:
+        grade_summary = grades.grade(student, request, course)
+
+    return success_cutoff and grade_summary['percent'] > success_cutoff
+
+
+def certificate_downloadable_status(student, course_key):
+    """
+    check the student existing certificates against a given course.
+    if status is not generating and not downloadable or error then user can view the generate button.
+
+    Args:
+        student : user object
+        course_key :  id associated with the course
+    Returns:
+        Dict containing student passed status also download url for cert if available
```

Comment:
Style nitpicks with this docstring:

1) Please capitalize the first sentence ("check the student).
2) Please add blank lines between the "Args" and "Returns" section.
3) Please remove the whitespace after `student` and `course_key` and add the types (so it's "student (User): user object" and "course_key (CourseKey): ID associated with the course")